internal/controller/clientpool/clientpool.go

```diff
@@ -131,8 +173,31 @@ func (cp *ClientPool) Close() {
 	defer cp.mux.Unlock()
 
 	for _, c := range cp.clients {
-		c.Close()
+		c.Client.Close()
+	}
+
+	cp.clients = make(map[ClientPoolKey]ClientInfo)
+}
+
+func (cp *ClientPool) isCertificateExpired(certBytes []byte) (bool, error) {
```

Comment:
I think it would be better to refresh the certificate _before_ it's expired not _when_ it's expired.

For example, the certificate we are working with in our cicd dogfooding expires after 24h, but I think it is renewed after 16h. This means that between 16h-24h, the new cert is available (I think).

Renewing the client _before_ it expires will avoid downtime if the controller needs to make calls right at the moment of expiry. An hour or 30 minutes before expiry is probably a safe time to refresh the client IMO.

If the user renews their cert only 5 minutes before expiry, then the controller would recreate the client unnecessarily a few times, but I think that's alright.